Element targeting via search text should resolve to elements with existing visibility adjustment

whsieh opened the pull request:
#### d9fadec5e2d75447f5cf247746a5d019b9a127f6
<pre>
Element targeting via search text should resolve to elements with existing visibility adjustment
<a href="https://bugs.webkit.org/show_bug.cgi?id=277932">https://bugs.webkit.org/show_bug.cgi?id=277932</a>
<a href="https://rdar.apple.com/131716603">rdar://131716603</a>

Reviewed by Abrar Rahman Protyasha.

Adjust element targeting codepaths such that targeted element requests that are based on search
text can resolve to elements that are inside of hidden subtrees (i.e. visibility adjustment).

* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::ClearVisibilityAdjustmentForScope::ClearVisibilityAdjustmentForScope):

Implement a move constructor so that I can construct a `Vector&lt;&gt;` of scopes below.

(WebCore::ClearVisibilityAdjustmentForScope::~ClearVisibilityAdjustmentForScope):

Pull this class up to the top of the file, so that we can use it in `findTargets` as well.

(WebCore::shouldIgnoreExistingVisibilityAdjustments):
(WebCore::ElementTargetingController::findTargets):

See above for more details.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
(TestWebKitAPI::TEST(ElementTargeting, RequestTargetedElementsBySearchableText)):

Augment an existing API test to exercise the change.

Canonical link: <a href="https://commits.webkit.org/282109@main">https://commits.webkit.org/282109@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de26e15e0e07a71ad8c13c645371734de3eedf8d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62108 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41462 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14700 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66088 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12653 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64227 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49148 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12993 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50025 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8746 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65177 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38488 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53790 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30852 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35137 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11030 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11584 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56922 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11334 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67816 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6051 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11098 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57402 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6077 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53743 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57649 "Found 1 new API test failure: /WebKitGTK/TestUIClient:/webkit/WebKitWebView/file-chooser-request (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13801 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4966 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37262 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38346 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39442 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38091 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -784,8 +813,22 @@ static inline std::optional<IntRect> inflatedClientRectForAdjustmentRegionTracki
+
 Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElementRequest&& request)
 {
+    Vector<std::unique_ptr<ClearVisibilityAdjustmentForScope>> clearVisibilityAdjustmentScopes;
```
This can just be a vector of values, right?

whsieh replied:
Hmm, I think it's possible but likely a bit ugly since this RAII class is non-copyable. I'll try it out and see...

Thanks for the review!

Edit: Nice, Just Works with `Vector` as long as I implement a move ctor.

whsieh committed:
Element targeting via search text should resolve to elements with existing visibility adjustment
https://bugs.webkit.org/show_bug.cgi?id=277932
rdar://131716603

Reviewed by Abrar Rahman Protyasha.

Adjust element targeting codepaths such that targeted element requests that are based on search
text can resolve to elements that are inside of hidden subtrees (i.e. visibility adjustment).

* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::ClearVisibilityAdjustmentForScope::ClearVisibilityAdjustmentForScope):

Implement a move constructor so that I can construct a `Vector<>` of scopes below.

(WebCore::ClearVisibilityAdjustmentForScope::~ClearVisibilityAdjustmentForScope):

Pull this class up to the top of the file, so that we can use it in `findTargets` as well.

(WebCore::shouldIgnoreExistingVisibilityAdjustments):
(WebCore::ElementTargetingController::findTargets):

See above for more details.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
(TestWebKitAPI::TEST(ElementTargeting, RequestTargetedElementsBySearchableText)):

Augment an existing API test to exercise the change.

Canonical link: https://commits.webkit.org/282109@main
